mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraThreadDAO.java

```diff
@@ -0,0 +1,112 @@
+import static org.apache.james.mailbox.cassandra.table.CassandraMessageIds.MESSAGE_ID;
+import static org.apache.james.mailbox.cassandra.table.CassandraThreadTable.BASE_SUBJECT;
+import static org.apache.james.mailbox.cassandra.table.CassandraThreadTable.MIME_MESSAGE_ID;
+import static org.apache.james.mailbox.cassandra.table.CassandraThreadTable.TABLE_NAME;
+import static org.apache.james.mailbox.cassandra.table.CassandraThreadTable.USERNAME;
+
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.core.Username;
+import org.apache.james.mailbox.cassandra.ids.CassandraMessageId;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.ThreadId;
+import org.apache.james.mailbox.store.mail.model.MimeMessageId;
+import org.apache.james.mailbox.store.mail.model.Subject;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+
+import reactor.core.publisher.Flux;
+
+public class CassandraThreadDAO {
+    private final CassandraAsyncExecutor executor;
+    private final PreparedStatement insertOne;
+    private final PreparedStatement selectOne;
+    private final PreparedStatement deleteOne;
+
+    @Inject
+    public CassandraThreadDAO(Session session) {
+        executor = new CassandraAsyncExecutor(session);
+
+        insertOne = session.prepare(insertInto(TABLE_NAME)
+            .value(USERNAME, bindMarker(USERNAME))
+            .value(MIME_MESSAGE_ID, bindMarker(MIME_MESSAGE_ID))
+            .value(MESSAGE_ID, bindMarker(MESSAGE_ID))
+            .value(THREAD_ID, bindMarker(THREAD_ID))
+            .value(BASE_SUBJECT, bindMarker(BASE_SUBJECT)));
+
+        selectOne = session.prepare(select(BASE_SUBJECT, THREAD_ID)
+            .from(TABLE_NAME)
+            .where(eq(USERNAME, bindMarker(USERNAME)))
+            .and(eq(MIME_MESSAGE_ID, bindMarker(MIME_MESSAGE_ID))));
+
+        deleteOne = session.prepare(delete().from(TABLE_NAME)
+            .where(eq(USERNAME, bindMarker(USERNAME)))
+            .and(eq(MIME_MESSAGE_ID, bindMarker(MIME_MESSAGE_ID))));
+    }
+
+    public Flux<Void> insertSome(Username username, Set<MimeMessageId> mimeMessageIds, MessageId messageId, ThreadId threadId, Optional<Subject> baseSubject) {
+        return Flux.fromIterable(mimeMessageIds)
+            .flatMap(mimeMessageId -> executor.executeVoid(insertOne.bind()
+                .setString(USERNAME, username.asString())
+                .setString(MIME_MESSAGE_ID, mimeMessageId.getValue())
+                .setUUID(MESSAGE_ID, ((CassandraMessageId) messageId).get())
+                .setUUID(THREAD_ID, ((CassandraMessageId) threadId.getBaseMessageId()).get())
+                .setString(BASE_SUBJECT, baseSubject.map(Subject::getValue).orElse(null))));
```

Comment:
Specify concurrency upon flatmap

Reply:
Do you mean I should put CassandraAsyncExecutor.execute... beyond?